### PR TITLE
Implement TrueType character map "format 2" (fixes #14117)

### PR DESCRIPTION
**Pull request by jberkenbilt**
If a PDF included an embedded TrueType font whose preferred character
map (cmap) was in "format 2", the code would select that character map
and then refuse to read it because of an unsupported format, thus
causing the characters not to be rendered. This commit implements
support for format 2 as described at the link below.

https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6cmap.html

### CODE REVIEW

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
> This lints cleanly and works on a small sample of my test files that I have exercised it with.

Did you manage to run the PDF.js test-suite successfully with these changes?
Please see https://github.com/mozilla/pdf.js/wiki/Contributing#4-run-lint-and-testing

---

Overall, this looks really good and seems to follow the specification nicely!
I've added a couple of comments, and please remember to squash the commits when addressing them.

File: src/core/base_stream.js [hunk 1]
```diff
@@ -67,6 +67,12 @@ class BaseStream {
     return (b0 << 8) + b1;
   }
 
+  getInt16() {
+    const b0 = this.getByte();
+    const b1 = this.getByte();
+    return (b0 << 8) + b1;
+  }
+
```
[the remark this answers is not in the document]
Comment (jberkenbilt) [reply src/core/base_stream.js:75]: It occurred to me overnight that this was wrong. I should do getUint16() and subtract 65536 if the result is > 32767.

[the remark this answers is not in the document]
Comment (Snuffleupagus) [reply src/core/base_stream.js:75]: > I should do getUint16() and subtract 65536 if the result is > 32767.

In that case, do you perhaps want to use this existing helper instead?
https://github.com/mozilla/pdf.js/blob/39459656052fe822bb7f57eccf8c029c804f64f7/src/core/fonts.js#L251-L254

To use that one, in the font code, you'd basically do something along these lines:
```javascript
const int = signedInt16(file.getByte(), file.getByte());
```

File: src/core/fonts.js [hunk 2]
```diff
@@ -1497,6 +1497,58 @@ class Font {
...
+            idDelta: file.getInt16(),
+            idRangePos: file.pos,
+          };
+          const idRangeOffset = file.getUint16();
+          subHeader.idRangePos += idRangeOffset;
+          subHeaders.push(subHeader);
+        }
+        for (let i = 0; i < 256; i++) {
+          if (subHeaderKeys[i] === 0) {
+            // i is a single-byte code
```
[the remark this answers is not in the document]
Comment (jberkenbilt) [reply src/core/fonts.js]: > Nit: Here, and below, please remember to add a period at the end of all comments.

i is adding periods to all my comments that are sentences

File: test/pdfs/.gitignore [hunk 3]
```diff
@@ -23,6 +23,7 @@
 !issue5256.pdf
 !issue5801.pdf
 !issue5946.pdf
+!issue14117-ttf-cmap-format-2.pdf
```
[the remark this answers is not in the document]
Comment (jberkenbilt) [reply test/pdfs/.gitignore]: ok

**Comment by jberkenbilt** [comment]
Everything checks out for me locally. I'm just looking at automated tests now.

**Comment by jberkenbilt** [comment]
@Snuffleupagus BTW, I greatly appreciate the quick review and feedback. Also Mr. Snuffleupagus was always one of my favorite Sesame Street characters, and it was a great day indeed when Big Bird was vindicated about his existence. Sadly I was already an adult by then and missed the actual event.

**Review by Snuffleupagus — APPROVED** [review]
r=me, looks great; thank you for the patch!

**Pull request merged**